### ISSUE TITLE
Constness of input parameters for CMSIS-DSP functions

### ISSUE THREAD

**Issue by escherstair** [ARM-software/CMSIS_5#349]
Some functions of CMSIS-DSP library take vectors as input parameters (passed by pointer to the first element).
Some of these functions don't modify at all these vectors and so it would be better if the type of the input parameters is `const _type_* pSrc`.
As an example `arm_max_xxx()`, `arm_mean_xxx()`, `arm_min_xxx()`, and so on.
For other functions it's not clear if they can work in place by design (and so input vector cannot be pointer-to-const) or not (and so input vector could be declared as pointer-to-const).
As an example,I refer to `arm_q31_to_float()` and so on.

Could you review the declarations of the functions in CMSIS-DSP library and set the constness when needed?

**Comment by llefaucheur** [comment]
Hi, the current direction is 1) keep API as-is for legacy of customer code investments (including Mathworks)    2) add blas API    3) prepare the path to Cortex-A/R devices with vector API with strides (for example arm_vabs_TYP (TYP *pA, s32 stridesA, real_TYP *pOut, s32 stridesOut, s32 N, ctrl *S), TYP in {q7,q15,q31, .. zq7,zq15,zq31(complex) , f8,f16,f32.. zf8,zf16,zf31..}). The *ctrl being there for future extensions (circular buffer addressing, ..).

**Comment by escherstair** [comment]
Hi @llefaucheur 
sorry but I don't understand your answer.
Could you clarify it, please?

**Comment by escherstair** [comment]
In `arm_max_q31(q31_t *pSrc, uint32_t blockSize, q31_t *pResult, uint32_t *pIndex)`, **`pSrc`** should be declared as `const q31_t*` too.

**Comment by anton-dirac** [comment]
Why was this reverted?

**Comment by JonatanAntoni** [comment]
Hi @anton-dirac,

the provided patch broke the library.

Cheers,
Jonatan

**Comment by anton-dirac** [comment]
Ok! That's a good reason to revert it :)

**Comment by escherstair** [comment]
Do you have news on this task?

**Comment by christophe0606** [comment]
@escherstair It has been progressively added to the library.
If you see missing cases, thanks to tell me. I'll improve it.

**Comment by christophe0606** [comment]
We have been continuously improving the library for `const` correctness.
I consider this task mainly done. I am not saying it is perfect and there may be few remaining cases where `const` is missing. In that case, new github issues will be created.
